Ihor dev - try that

shybzzz opened the pull request:
this pull request is created to see whole bunch off changes before refactoring

ihor96berizka committed:
Merge remote-tracking branch 'origin/master' into ihor_dev

ihor96berizka committed:
Added expSmooth filter

ihor96berizka committed:
refactored data routine.

ihor96berizka committed:
Added newMean filter.

shybzzz commented on routine-data.h:
```diff
@@ -2,26 +2,27 @@
 
 using getData = int();
+using Filter = int();
 
```
we do not need Filter callback
letus do following
1. keep last 10 (DATA_SIZE) samples in memory queue
2. if new sample comes push back it and pop front 1-st of the saved measurements. this is how data queue is up to date. do this in gatherData function.
3. calculate values with ALL of the filters on the updated data and publish each into its mqtt topic. letus have "filter/median", "filter/mean" and "filter/exp_smooth" topics for filtered values and "data" for raw ,easurement (the one we've just pushed).

this is how I would desire this functionality to be refactored
and this is why we do not need filter callback - just call all of them

ihor96berizka replied:
I guess. that we will need filter callback here.

Without callback - we will need to publish filtered data in processData function.
Or - we can return tuple\map of filtered values from this function(This approach is not good. cause it assumes using of another container from c++ standard library).

And the best approach(in my opinion) is to use callback.
We will pass callbacks to processData function. Fetch results and then publish data in runDataRoutine function.

shybzzz commented on init-data.h:
```diff
@@ -3,7 +3,10 @@
+#define MEDIAN_MID MEDIAN_WINDOW/2
+#define EXP_SMOOTH_WINDOW 1u
+#define NEW_MEAN_WINDOW 1u
 
```
why do we need all of these windows?

can we have only DATA_SIZE and work with it in all filters?

ihor96berizka replied:
This windows are defined for each filter.
Cause we need only one raw data sample - window sizes are equal to 1.(for expSmoth and newMean filters).
I guess that this variables can be removed if I implement first comment.

shybzzz commented on filter-new-mean.h:
```diff
@@ -0,0 +1,20 @@
+#define __FILTER_NEW_MEAN__
+
+int filterNewMean()
+{
```
I do not understand how this filter works
shouldnt that be just sum(data)/DATA_SIZE

ihor96berizka replied:
![newmean](https://user-images.githubusercontent.com/18075031/49136805-70e79500-f2f3-11e8-9d60-3369af558e2b.PNG)

shybzzz replied:
got it

shybzzz commented on filter-expSmooth.h:
```diff
@@ -0,0 +1,20 @@
+{
+  double alp = 0.5;
+  static int prev_filtered_sample = -1;
+  if (prev_filtered_sample < 0)
```
this is strange
letus avoid static variables - it is not straightforward enough what is the value of each
this can easily be avoided by passing prev va;ue in argument ar smth like that

ihor96berizka committed:
Fixed comments and added logs for debugging.

Need to think about using #ifdef and global variable for logs management. For example if set some value - print all logs.
